src/main/kotlin/registry/K8sRegistry.kt

```diff
@@ -0,0 +1,142 @@
+package leia.registry
+
+import com.fasterxml.jackson.core.JsonFactory
+import com.fasterxml.jackson.databind.DeserializationFeature
+import com.fasterxml.jackson.databind.ObjectMapper
+import com.fasterxml.jackson.module.kotlin.KotlinModule
+import com.fasterxml.jackson.module.kotlin.MissingKotlinParameterException
+import io.ktor.client.HttpClient
+import io.ktor.client.call.call
+import io.ktor.client.request.HttpRequestBuilder
+import io.ktor.client.response.readBytes
+import io.ktor.http.HttpMethod
+import io.ktor.http.HttpStatusCode
+import kotlinx.coroutines.experimental.runBlocking
+import mu.KLogging
+import java.net.ConnectException
+import java.nio.channels.UnresolvedAddressException
+import java.util.concurrent.Executors
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.atomic.AtomicReference
+
+class K8sResourceHolder(private val parser: (String) -> List<K8sRegistry.RouteItem>) {
+    private val entriesA = AtomicReference(listOf<K8sRegistry.RouteItem>())
+    fun onChange(yaml: String) {
+        entriesA.updateAndGet {
+            try {
+                parser(yaml)
+            } catch (e: MissingKotlinParameterException) {
+                K8sRegistry.logger.error { K8sRegistry.logger.error { "Failed to read objects from kubernetes: ${e.message}" } }
+                it
+            }
+        }
+    }
+
+    fun getData(): List<K8sRegistry.RouteItem> = entriesA.get()
+}
+
+// Auto-watching registry for a directory of Kubernetes Yaml files.
+class K8sRegistry(private val host: String, private val port: String) : Registry {
+    private val watchers = mutableListOf<Triple<String, (Map<String, Any>) -> Any, (List<*>) -> Unit>>()
+    private val holder = K8sResourceHolder { yaml ->
+        val mapper = ObjectMapper(JsonFactory())
```

Comment:
Is this called multiple times if so we could possibly move the mapper init into the class

Reply:
Done.